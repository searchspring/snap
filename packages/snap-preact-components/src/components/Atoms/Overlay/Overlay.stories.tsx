import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY, Markdown } from '@storybook/blocks';

import { Overlay, OverlayProps } from './Overlay';
import { componentArgs, highlightedCode } from '../../../utilities';
import Readme from '../Overlay/readme.md';

export default {
	title: 'Atoms/Overlay',
	component: Overlay,
	tags: ['autodocs'],
	parameters: {
		docs: {
			page: () => (
				<div>
					<Markdown
						options={{
							overrides: {
								code: highlightedCode,
							},
						}}
					>
						{Readme}
					</Markdown>
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	argTypes: {
		active: {
			description: 'Overlay is displayed',
			type: { required: true },
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: {
				type: 'boolean',
			},
		},
		color: {
			defaultValue: 'rgba(0,0,0,0.8)',
			description: 'Overlay color',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: 'rgba(0,0,0,0.8)' },
			},
			control: { type: 'color' },
		},
		transitionSpeed: {
			defaultValue: '0.25s',
			description: 'Overlay opening/closing transition speed',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: '0.25s' },
			},
			control: { type: 'text' },
		},
		onClick: {
			description: 'Overlay click event handler',
			table: {
				type: {
					summary: 'function',
				},
			},
			control: { type: 'none' },
			action: 'onClick',
		},
		...componentArgs,
	},
};

export const Default = (args: OverlayProps) => <Overlay {...args} />;
Default.args = {
	active: true,
};
