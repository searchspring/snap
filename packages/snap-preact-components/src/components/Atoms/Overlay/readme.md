# Overlay

Renders an Overlay. Typically used to blur the background with a foreground element such as a modal or slideout menu.

## Usage

### active
The required `active` prop specifies when to render the component.

```jsx
<Overlay active={true} />
```

### color
The `color` prop specifies the color of the overlay.

```jsx
<Overlay active={true} color={'rgba(0,0,0,0.8)'} />
```

### transitionSpeed
The `transitionSpeed` prop specifies animation transition speed.

```jsx
<Overlay active={true} transitionSpeed='0.5s' />
```

### Events

#### onClick
The `onClick` prop allows for a custom callback function.

```jsx
<Overlay active={true} onClick={(e)=>{console.log(e)}} />
```
