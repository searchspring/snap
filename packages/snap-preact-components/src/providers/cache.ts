export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
