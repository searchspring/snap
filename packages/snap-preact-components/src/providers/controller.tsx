import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
