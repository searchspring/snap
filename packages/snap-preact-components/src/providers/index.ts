export * from './store';
export * from './theme';
