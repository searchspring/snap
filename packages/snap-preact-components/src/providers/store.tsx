import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
