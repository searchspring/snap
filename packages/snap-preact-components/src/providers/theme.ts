import { ThemeProvider as EmotionTheme } from '@emotion/react';
export { css, useTheme, withTheme } from '@emotion/react';

interface ThemeProviderProps {
	theme: Partial<Theme> | ((outerTheme: Theme) => Theme);
	children?: any;
}
interface ThemeProvider {
	(props: ThemeProviderProps): React.ReactElement;
}
export const ThemeProvider: ThemeProvider = EmotionTheme;

export const defaultTheme: Theme = {
	colors: {
		primary: '#3A23AD',
		secondary: '#00cee1',
		hover: '#f8f6fd',
		text: {
			secondary: '#ffffff',
		},
		message: {
			error: '#cc1212',
			warning: '#ecaa15',
			info: '#4c3ce2',
		},
	},
	components: {},
};
export interface Theme {
	variables?: {
		[key: string]: unknown;
	};
	colors?: {
		primary: string;
		secondary: string;
		hover?: string;
		text?: {
			secondary: string;
		};
		message?: {
			error: string;
			warning: string;
			info: string;
		};
	};
	components?: any;
	namedComponents?: any;
}
