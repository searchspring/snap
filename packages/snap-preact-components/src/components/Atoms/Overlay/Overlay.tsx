/** @jsx jsx */
import { h } from 'preact';

import { jsx, css } from '@emotion/react';
import classnames from 'classnames';

import { Theme, useTheme, CacheProvider } from '../../../providers';
import { ComponentProps } from '../../../types';

const CSS = {
	overlay: ({ color, transitionSpeed }) =>
		//@ts-ignore
		css({
			transition: `background ${transitionSpeed} ease 0s, left 0s ease ${transitionSpeed}`,
			position: 'fixed',
			zIndex: '10003',
			height: '100%',
			width: '100%',
			top: '0',
			left: '-100%',
			'&.ss__overlay--active': {
				transition: `background ${transitionSpeed} ease, left 0s ease`,
				background: color,
				left: '0',
			},
		}),
};

export function Overlay(properties: OverlayProps): JSX.Element {
	const globalTheme: Theme = useTheme();

	const props: OverlayProps = {
		// default props
		color: 'rgba(0,0,0,0.8)',
		transitionSpeed: '0.25s',
		// global theme
		...globalTheme?.components?.overlay,
		// props
		...properties,
		...properties.theme?.components?.overlay,
	};

	const { active, color, transitionSpeed, onClick, disableStyles, className, style } = props;

	const styling: { css?: any } = {};
	if (!disableStyles) {
		styling.css = [CSS.overlay({ color, transitionSpeed }), style];
	} else if (style) {
		styling.css = [style];
	}
	return (
		<CacheProvider>
			<div
				onClick={(e: React.MouseEvent<HTMLDivElement, Event>) => onClick && active && onClick(e)}
				className={classnames('ss__overlay', { 'ss__overlay--active': active }, className)}
				{...styling}
			/>
		</CacheProvider>
	);
}

export interface OverlayProps extends ComponentProps {
	active: boolean;
	color?: string;
	transitionSpeed?: string;
	onClick?: (e: React.MouseEvent<HTMLDivElement, Event>) => void;
}
