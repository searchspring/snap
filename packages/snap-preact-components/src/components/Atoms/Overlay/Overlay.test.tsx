import { h } from 'preact';

import { render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';

import { Overlay } from './Overlay';
import { ThemeProvider } from '../../../providers';

describe('Overlay Component', () => {
	const theme = {
		components: {
			overlay: {
				color: 'rgba(200, 100, 50, 0.5)',
			},
		},
	};

	it('renders', () => {
		const args = { active: true };
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		expect(overlayElement).toBeInTheDocument();
	});

	it('is hidden when inactive', () => {
		const args = { active: false };
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay.ss__overlay--active');
		expect(overlayElement).not.toBeInTheDocument();
	});

	it('is has custom color', () => {
		const args = {
			active: true,
			color: 'rgba(0, 0, 0, 0.3)',
		};
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		const styles = getComputedStyle(overlayElement);
		expect(styles.backgroundColor).toBe(args.color);
	});

	it('can disable styling', () => {
		const args = {
			active: true,
			disableStyles: true,
		};
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		const styles = getComputedStyle(overlayElement);
		expect(styles.backgroundColor).toBe('');
		expect(overlayElement.classList.length).toBe(3);
		expect(overlayElement.classList).toContain('css-0');
	});

	it('can custom className', () => {
		const args = {
			active: true,
			className: 'custom-class',
		};
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		expect(overlayElement.classList).toContain(args.className);
	});

	it('fires onClick prop when clicked', () => {
		const args = {
			active: true,
			onClick: jest.fn(),
		};
		const rendered = render(<Overlay {...args} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		userEvent.click(overlayElement);
		expect(args.onClick).toHaveBeenCalled();
	});

	it('is themeable with ThemeProvider', () => {
		const args = { active: true };
		const rendered = render(
			<ThemeProvider theme={theme}>
				<Overlay {...args} />
			</ThemeProvider>
		);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		const styles = getComputedStyle(overlayElement);
		expect(styles.backgroundColor).toBe(theme.components.overlay.color);
	});

	it('is themeable with theme prop', () => {
		const args = { active: true };
		const rendered = render(<Overlay {...args} theme={theme} />);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		const styles = getComputedStyle(overlayElement);
		expect(styles.backgroundColor).toBe(theme.components.overlay.color);
	});

	it('is themeable with theme prop overrides ThemeProvider', () => {
		const args = { active: true };
		const themeOverride = {
			components: {
				overlay: {
					color: 'rgba(100, 100, 100, 0.9)',
				},
			},
		};
		const rendered = render(
			<ThemeProvider theme={theme}>
				<Overlay {...args} theme={themeOverride} />
			</ThemeProvider>
		);
		const overlayElement = rendered.container.querySelector('.ss__overlay');
		const styles = getComputedStyle(overlayElement);
		expect(styles.backgroundColor).toBe(themeOverride.components.overlay.color);
		expect(styles.backgroundColor).not.toBe(theme.components.overlay.color);
	});
});
