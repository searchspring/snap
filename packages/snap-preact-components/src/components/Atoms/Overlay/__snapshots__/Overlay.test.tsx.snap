// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay Component can custom className 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active custom-class emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component can disable styling 1`] = `
<DocumentFragment>
  <div
    class="ss__overlay ss__overlay--active"
  />
</DocumentFragment>
`;

exports[`Overlay Component fires onClick prop when clicked 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component fires onClick prop when clicked 2`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component is has custom color & transition speed 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.35s ease 0s,left 0s ease 0.35s;
  transition: background 0.35s ease 0s,left 0s ease 0.35s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.35s ease,left 0s ease;
  transition: background 0.35s ease,left 0s ease;
  background: rgba(0, 0, 0, 0.3);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component is hidden when inactive 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(200, 100, 50, 0.5);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component is themeable with theme prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(200, 100, 50, 0.5);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component is themeable with theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(100, 100, 100, 0.9);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component renders 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;

exports[`Overlay Component uses bocachica theme 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transition: background 0.25s ease 0s,left 0s ease 0.25s;
  transition: background 0.25s ease 0s,left 0s ease 0.25s;
  position: fixed;
  z-index: 10003;
  height: 100%;
  width: 100%;
  top: 0;
  left: -100%;
}

.emotion-0.ss__overlay--active {
  -webkit-transition: background 0.25s ease,left 0s ease;
  transition: background 0.25s ease,left 0s ease;
  background: rgba(0,0,0,0.8);
  left: 0;
}

<div
    class="ss__overlay ss__overlay--active emotion-0"
  />
</DocumentFragment>
`;
